File: views/fun/sex.py
import random
import discord
from ext import update_quest

async def offer_sex(inter: discord.Interaction, user: discord.User):
    if user.bot: return await inter.response.send_message("Зачем ебать бота?", ephemeral=True)
    if user == inter.user: return await inter.response.send_message("Ты че ебать себя собираешься?", ephemeral=True)

    await inter.response.send_message(
        embed=discord.Embed(
            title=f"{user.global_name}, {inter.user.global_name} предложил Вам секс, Вы согласны?",
            color=discord.Color.random()
        ),
        view=SexView(inter.user, user)
    )

class SexView(discord.ui.View):
    def __init__(self, author: discord.User, user: discord.User):
        super().__init__(timeout=None)
        self.author = author
        self.user = user

    @discord.ui.button(label="Да", style=discord.ButtonStyle.success)
    async def yessex(self, inter: discord.Interaction, button: discord.ui.Button):
        if inter.user.id != self.user.id: return await inter.response.send_message("Завидуй молча, это не тебе секс предлагали", ephemeral=True)
        
        giffs = ["https://media.tenor.com/pn5xTq0WtqcAAAAC/anime-girl.gif", "https://media.tenor.com/9G1zsVIiV6UAAAAC/anime-bed.gif", "https://media.tenor.com/tdK59AzAWZgAAAAC/pokemon-anime.gif"
                    , "https://media.tenor.com/i7S2Taae5H8AAAAC/sex-anime.gif", "https://media.tenor.com/eq-B2_glw0sAAAAC/ver-anime.gif"]
        randgif = random.choice(giffs)
        soglaz=discord.Embed(title=f"**{self.user.global_name} согласился на секс с {self.author.global_name}**", color=discord.Color.random())
        soglaz.set_image(url=randgif)

        for x in self.children:
            x.disabled = True

        await inter.response.edit_message(view=self)
        await inter.followup.send(embed=soglaz)

        await update_quest(self.author, "sex", used_user=self.user)

    @discord.ui.button(label="Нет", style=discord.ButtonStyle.danger)
    async def nosex(self , inter: discord.Interaction, button: discord.ui.Button):
        if inter.user.id != self.user.id: return await inter.response.send_message("Завидуй молча, это не тебе секс предлагали", ephemeral=True)
        
        for x in self.children:
            x.disabled = True

        await inter.response.edit_message(view=self)
        await inter.followup.send(f"**{self.author.mention}, вот чёрт, тебе отказал {self.user.mention} :(**")

File: src/modules/fun/sex.rs
use crate::buttons::handle_buttons;
use crate::types::*;
use poise::CreateReply;
use poise::serenity_prelude::Mentionable;
use rand::seq::IndexedRandom;

const GIFS: [&str; 5] = [
    "https://media.tenor.com/pn5xTq0WtqcAAAAC/anime-girl.gif",
    "https://media.tenor.com/9G1zsVIiV6UAAAAC/anime-bed.gif",
    "https://media.tenor.com/tdK59AzAWZgAAAAC/pokemon-anime.gif",
    "https://media.tenor.com/i7S2Taae5H8AAAAC/sex-anime.gif",
    "https://media.tenor.com/eq-B2_glw0sAAAAC/ver-anime.gif",
];

/// Предложить секс пользователю
#[poise::command(
    slash_command,
    context_menu_command = "Предложить секс",
    install_context = "User | Guild",
    interaction_context = "Guild | BotDm | PrivateChannel"
)]
pub async fn sex(
    ctx: Context<'_>,
    #[description = "Кому предлагается секс"] user: serenity::User,
) -> Result<(), Error> {
    if user.bot {
        ctx.send(CreateReply::default().content("Роботофил!").ephemeral(true))
            .await?;
        return Ok(());
    }

    if user.id == ctx.author().id {
        ctx.send(
            CreateReply::default()
                .content("Ты чё ебать себя собираешься?")
                .ephemeral(true),
        )
        .await?;
        return Ok(());
    }

    let mut buttons = vec![serenity::CreateActionRow::Buttons(vec![
        serenity::CreateButton::new(format!("{}:sex:yes", ctx.id()))
            .label("Да")
            .style(serenity::ButtonStyle::Success),
        serenity::CreateButton::new(format!("{}:sex:no", ctx.id()))
            .label("Нет")
            .style(serenity::ButtonStyle::Danger),
    ])];

    let embed = serenity::CreateEmbed::default()
        .title(format!(
            "{}, {} предложил Вам секс, Вы согласны?",
            user.global_name.as_deref().unwrap_or_else(|| &user.name),
            ctx.author()
                .global_name
                .as_deref()
                .unwrap_or_else(|| &ctx.author().name)
        ))
        .colour(serenity::colours::branding::GREEN);

    let msg = ctx
        .send(
            CreateReply::default()
                .embed(embed)
                .components(buttons.clone()),
        )
        .await?;

    for row in &mut buttons {
        if let serenity::CreateActionRow::Buttons(btns) = row {
            for button in btns {
                *button = button.clone().disabled(true);
            }
        }
    }

    let press_user = user.clone();
    handle_buttons(
        ctx,
        format!("{}:sex:", ctx.id()).as_str(),
        300,
        move |press, relative_id| {
            let user = press_user.clone();
            let buttons = buttons.clone();

            async move {
                if press.user.id != user.id {
                    crate::create_response!(
                        ctx,
                        press,
                        serenity::CreateInteractionResponseMessage::default()
                            .content("Завидуй молча, это не тебе секс предлагали")
                            .ephemeral(true)
                    );
                    return Ok(false);
                }

                if relative_id == "yes" {
                    crate::create_edit_response!(
                        ctx,
                        press,
                        serenity::CreateInteractionResponseMessage::default().components(buttons)
                    );

                    let gif = {
                        let mut rng = rand::rng();
                        *GIFS.choose(&mut rng).unwrap()
                    };

                    let embed = serenity::CreateEmbed::default()
                        .title(format!(
                            "{} согласился на секс с {}",
                            user.global_name.as_deref().unwrap_or_else(|| &user.name),
                            ctx.author()
                                .global_name
                                .as_deref()
                                .unwrap_or_else(|| &ctx.author().name)
                        ))
                        .image(gif)
                        .colour(serenity::colours::branding::GREEN);

                    press
                        .create_followup(
                            &ctx,
                            serenity::CreateInteractionResponseFollowup::new().embed(embed),
                        )
                        .await?;
                } else if relative_id == "no" {
                    crate::create_edit_response!(
                        ctx,
                        press,
                        serenity::CreateInteractionResponseMessage::default().components(buttons)
                    );

                    press
                        .create_followup(
                            &ctx,
                            serenity::CreateInteractionResponseFollowup::new().content(format!(
                                "**{}, вот чёрт, тебе отказал {} :(**",
                                ctx.author().mention(),
                                user.mention()
                            )),
                        )
                        .await?;
                }
                Ok(true)
            }
        },
        move || async move {
            let embed = serenity::CreateEmbed::default()
                .title(format!(
                    "{} не успел согласиться на предложение секса от {}",
                    user.global_name.as_deref().unwrap_or_else(|| &user.name),
                    ctx.author()
                        .global_name
                        .as_deref()
                        .unwrap_or_else(|| &ctx.author().name)
                ))
                .colour(serenity::colours::branding::RED);

            msg.edit(
                ctx,
                CreateReply::default().components(Vec::new()).embed(embed),
            )
            .await?;
            Ok(())
        },
    )
    .await?;

    Ok(())
}
